# Numerical Methods 1
### [Gerard Gorman](http://www.imperial.ac.uk/people/g.gorman), [Matthew Piggott](http://www.imperial.ac.uk/people/m.d.piggott), [Nicolas Barral](http://www.imperial.ac.uk/people/n.barral)

# Lecture 5: Numerical Linear Algebra I

## Learning objectives:

* Manipulation of matrices and matrix equations in Python
* Reminder on properties of matrices (from MM1): determinants, singularity etc
* Algorithms for the solution of linear systems (this week we focus on so-called direct methods)
* Gaussian elimination, including back substitution

## Introduction - Linear (matrix) systems

Recall from your Mathematical Methods I course that the we can re-write a system of simultaneous (linear) equations in matrix form.  For example, in week 4 of MM1 you considered the following example:

\begin{eqnarray*}
  2x + 3y &=& 7 \\
   x - 4y &=& 3
\end{eqnarray*} 

and it was noted that this can be written in matrix form as 

$$
\left(
  \begin{array}{rr}
    2 & 3 \\
    1 & -4  \\
  \end{array}
\right)\left(
  \begin{array}{c}
    x \\
    y \\
  \end{array}
\right) = \left(
  \begin{array}{c}
    7 \\
    3 \\
  \end{array}
\right)
$$

More generally, consider the arbitrary system of $n$ linear equations for $n$ unknowns

\begin{eqnarray*}
  A_{11}x_1 + A_{12}x_2 + \dots + A_{1n}x_n &=& b_1 \\ 
  A_{21}x_1 + A_{22}x_2 + \dots + A_{2n}x_n &=& b_2 \\ 
  \vdots &=& \vdots \\ 
  A_{n1}x_1 + A_{n2}x_2 + \dots + A_{nn}x_n &=& b_n
\end{eqnarray*}

where $A_{ij}$ are the constant coefficients of the linear system, $x_j$ are the unknown variables, and $b_i$
are the terms on the right hand side (RHS).  Here the index $i$ is referring to the equation number
(the row in the matrix below), with the index $j$ referring to the component of the unknown
vector $\pmb{x}$ (the column of the matrix).

This system of equations can be represented as the matrix equation $A\pmb{x}=\pmb{b}$: 

$$
\left(
  \begin{array}{cccc}
    A_{11} & A_{12} & \dots & A_{1n} \\
    A_{21} & A_{22} & \dots & A_{2n} \\
    \vdots & \vdots & \ddots & \vdots \\
    A_{n1} & A_{n2} & \dots & A_{nn} \\
  \end{array}
\right)\left(
         \begin{array}{c}
           x_1 \\
           x_2 \\
           \vdots \\
           x_n \\
         \end{array}
       \right)  = \left(
                   \begin{array}{c}
                     b_1 \\
                     b_2 \\
                     \vdots \\
                     b_n \\
                   \end{array}
                 \right)
$$


We solved the above $2 \times 2$ example of two equations and two unknowns in MM1 using substitution (*i.e.* we multiplied the second equation by 2 and subtracted the first equation from the resulting equation to eliminate $x$ and hence allowing us to find $y$, then we could compute $x$ from the first equation), so that we could easily find the solution 

$$ x=37/11, \quad y=1/11.$$

In MM1 you also considered $3 \times 3$ examples which were a little more complicated but still doable.  This lecture considers the case of $n \times n$ where $n$ could easily be billions! This case arises when you solve a differential equation numerically on a discrete mesh or grid. Here you would typically obtain one unknown and one (discrete, linear or nonlinear) equation at very grid point. You could generate an arbitrarily large matrix system simply by generating a finer mesh.

Note that you will solve differential equations numerically in the follow-up course Numerical Methods II.

Cases where the matrix is non-square, i.e. of shape $m \times n$ where $m\ne n$ correspond to the (over- or under-determined) system where you have more or less equations than unknowns - we won't consider these in this lecture. 


## Matrices in Python

We have already used numpy arrays to store one-dimensional vectors of numbers.

The convention is that these are generally considered to be *column* vectors and have shape $n \times 1$.

We can extend to higher dimensions through the introduction of matrices as two-dimensional arrays (more generally vectors and matrices are just two examples of tensors). 

We use subscript indices to identify each component of the array or matrix, i.e. we can identify each component of the vector $\pmb{v}$ by $v_i$, and each component of the matrix $A$ by $A_{ij}$.  

Note that it is a convention that vectors are either underlined or bold, and generally lower case letters, whereas matrices are plain capital letters.

The *dimension* or *shape* of a vector/matrix is the number of rows and columns it posesses, i.e. $n \times 1$ and $m \times n$ for the examples above.

Here is an example of how we can extend our use of the numpy array object to two dimensions in order to define a matrix $A$ and some examples of some operations we can make on it.

In [1]:
import numpy
from scipy import linalg
A=numpy.array([[10., 2., 1.],[6., 5., 4.],[1., 4., 7.]])

print(A)

[[ 10.   2.   1.]
 [  6.   5.   4.]
 [  1.   4.   7.]]


In [2]:
# the total size of the array storing A - here 9 for a 3x3 matrix
print(numpy.size(A))  

9


In [3]:
# the number of dimensions of the matrix A
print(numpy.ndim(A) )    

2


In [4]:
# the shape of the matrix A
print(numpy.shape(A))

(3, 3)


In [5]:
# the transpose of the matrix A
print(A.T)

[[ 10.   6.   1.]
 [  2.   5.   4.]
 [  1.   4.   7.]]


In [6]:
# the inverse of the matrix A - computed using a scipy algorithm
print(linalg.inv(A))

[[ 0.14285714 -0.07518797  0.02255639]
 [-0.28571429  0.51879699 -0.2556391 ]
 [ 0.14285714 -0.28571429  0.28571429]]


In [7]:
# the determinant of the matrix A - computed using a scipy algorithm
print(linalg.det(A))

133.00000000000003


In [8]:
# Multiply A with its inverse using numpy's dot. Note that due to
# roundoff errors the off diagonal values are not exactly zero.
print(numpy.dot(A,linalg.inv(A)))

[[  1.00000000e+00   1.11022302e-16  -5.55111512e-17]
 [ -4.44089210e-16   1.00000000e+00   0.00000000e+00]
 [ -4.44089210e-16   6.66133815e-16   1.00000000e+00]]


In [9]:
# same way to achieve the same thing
print(A.dot(linalg.inv(A)))

[[  1.00000000e+00   1.11022302e-16  -5.55111512e-17]
 [ -4.44089210e-16   1.00000000e+00   0.00000000e+00]
 [ -4.44089210e-16   6.66133815e-16   1.00000000e+00]]


In [10]:
# note that the * operator simply does operations element-wise
print(A*linalg.inv(A))

[[ 1.42857143 -0.15037594  0.02255639]
 [-1.71428571  2.59398496 -1.02255639]
 [ 0.14285714 -1.14285714  2.        ]]


In [11]:
# how to initialise a vector of zeros 
print(numpy.zeros(3))

[ 0.  0.  0.]


In [12]:
# how to initialise a matrix of zeros 
print(numpy.zeros((3,3)))

[[ 0.  0.  0.]
 [ 0.  0.  0.]
 [ 0.  0.  0.]]


In [13]:
# how to initialise a 3rd-order tensor of zeros
print(numpy.zeros((3,3,3)))

[[[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]

 [[ 0.  0.  0.]
  [ 0.  0.  0.]
  [ 0.  0.  0.]]]


In [14]:
# how to initialise the identity matrix, I or Id
print(numpy.eye(3))

[[ 1.  0.  0.]
 [ 0.  1.  0.]
 [ 0.  0.  1.]]


Let's quickly consider the $2 \times 2$ case from MM1 recreated above where we claimed that $x=37/11$ and $y=1/11$.  

To solve the matrix equation 

$$ A\pmb{x}=\pmb{b} $$

we can simply multiply both sides by the inverse of the matrix $A$ (if $A$ is invertible and if we know what the inverse is of course!):

\begin{align}
A\pmb{x} & = \pmb{b}\\
\implies A^{-1}A\pmb{x} & = A^{-1}\pmb{b}\\
\implies I\pmb{x} & = A^{-1}\pmb{b}\\
\implies \pmb{x} & = A^{-1}\pmb{b}
\end{align}

so we can find the solution $\pmb{x}$ by multiplying the inverse of $A$ with the RHS vector $\pmb{b}$.

In [15]:
A=numpy.array([[2., 3.],[1., -4.]])

# check first whether the determinant of A is non-zero - see below for reasons why.
print(linalg.det(A))   

-11.0


In [16]:
b=numpy.array([7., 3.])

# compute A inverse and multiply by b
print(numpy.dot(linalg.inv(A),b))

[ 3.36363636  0.09090909]


In [17]:
# or solve the linear system using scipy - actually does 
# the same thing as line above using LU (see below)
print(linalg.solve(A,b))

print("Check it against the solution we calculated by hand earlier: ", 37./11., 1./11.)

[ 3.36363636  0.09090909]
Check it against the solution we calculated by hand earlier:  3.3636363636363638 0.09090909090909091


### <span style="color:blue">Aside: matrix objects </span>

Note that numpy does possess a matrix object as a sub-class of the numpy array.  We can cast the above two-dimensional arrays into matrix objects and then the star operator does yield the expected matrix product:

In [18]:
# the is an n-dimensional array (n-2 here)
print(type(A))

<class 'numpy.ndarray'>


In [19]:
# this casts the array A into the matrix class
print(type(numpy.mat(A)))

<class 'numpy.matrixlib.defmatrix.matrix'>


In [20]:
# for these objects * is standard matrix multuiplication 
# and we can check that A*A^{-1}=I as expected
print(numpy.mat(A)*numpy.mat(linalg.inv(A)))

[[ 1.  0.]
 [ 0.  1.]]


### <span style="color:blue">Slicing </span>

Remember from last term that just as for arrays or lists, we can use *slicing*  in order to extract components of matrices, for example:

In [21]:
A=numpy.array([[10., 2., 1.],[6., 5., 4.],[1., 4., 7.]])
print(A)

[[ 10.   2.   1.]
 [  6.   5.   4.]
 [  1.   4.   7.]]


In [22]:
# single entry, first row, second column
print(A[0,1])

2.0


In [23]:
# first row
print(A[0,:])

[ 10.   2.   1.]


In [24]:
# last row
print(A[-1,:])

[ 1.  4.  7.]


In [25]:
# second column
print(A[:,1])

[ 2.  5.  4.]


In [26]:
# extract a 2x2 sub-matrix
print(A[1:3,1:3])

[[ 5.  4.]
 [ 4.  7.]]


### <span style="color:blue">Exercise 5.1: Matrix manipulation in python </span>

Let
$$
A = \left(
  \begin{array}{ccc}
    1 & 2 & 3 \\
    4 & 5 & 6 \\
    7 & 8 & 9 \\
  \end{array}
\right)
\mathrm{\quad\quad and \quad\quad}
b = \left(
  \begin{array}{c}
    2 \\
    4 \\
    6 \\
  \end{array}
\right)
$$

- Store $A$ and $b$ in adapted python structures. Print them.
- Print their shape and size. What is the difference ?
- Create a python matrix I containing the identity matrix $I_3$. Do $A = A+I$.
- Substitute the 3rd column of $A$ with $b$.
- Solve $Ax=b$.

## Properties of matrices: determinants, singularity, solvability of linear systems, etc

Consider $N$ linear equations in $N$ unknowns, $A\pmb{x}=\pmb{b}$. 

From MM1 you learnt that this system has a *unique solution* provided that the determinant of A, $\det(A)$, is non-zero. In this case the matrix is said to be *non-singular*.

If $\det(A)=0$ (with $A$ then termed a *singular matrix*), then the linear system does *not* have a unique solution, it may have either infinite *or* no solutions.

For example, consider

$$
\left(
  \begin{array}{rr}
    2 & 3 \\
    4 & 6  \\
  \end{array}
\right)\left(
  \begin{array}{c}
    x \\
    y \\
  \end{array}
\right) = \left(
  \begin{array}{c}
    4 \\
    8 \\
  \end{array}
\right)
$$

The second equation is simply twice the first, and hence a solution to the first equation is also automatically a solution to the second equation. 

We hence only have one *linearly-independent* equation, and our problem is under-constrained: we effectively only have one eqution for two unknowns with infinitely many possibly solutions. 

If we replaced the RHS vector with $(4,7)^T$, then the two equations would be contradictory: in this case we have no solutions.

Note that a set of vectors where one can be written as a linear sum of the others are termed *linearly-dependent*. When this is not the case the vectors are termed *linearly-independent*.

The following properties of a square $n\times n$ matrix are equivalent:

* $\det(A)\ne 0$ - A is non-singular

* The columns of $A$ are linearly independent

* The rows of $A$ are linearly independent

* The columns of A *span* $n$-dimensional space (recall MM1 - we can reach any point in $\mathbb{R}^N$ through a linear combination of these vectors - note that this is simply what the operation $A\pmb{x}$ is doing of course if you write it out)

* $A$ is invertible, i.e. there exists a matrix $A^{-1}$ such that $A^{-1}A = A A^{-1}=I$

* the matrix system $A\pmb{x}=\pmb{b}$ has a unique solution for every vector $b$


## Gaussian elimination - method


The Gaussian elimination algorithm is simply a systematic implementation of the method of equation substitution we used above to solve the $2\times 2$ system (i.e. where we "multiply the second equation by 2 and subtract the first equation from the resulting equation to *eliminate* $x$ and hence allowing us to find $y$, we can then compute $x$ from the first equation").

So *Gaussian elimination*  as the method is atributed to the mathematician *Gauss* (although it was certainly known before his time) and *elimination* as we seek to eliminate unknowns.

To perform this method for arbitrarily large systems (on paper) we form the so-called augmented matrix

$$
[A|\pmb{b}] = 
\left[
  \begin{array}{rr|r}
    2 & 3 & 7 \\
    1 & -4 & 3  \\
  \end{array}
\right]
$$

First we multiplied the second equation by 2, this yield the updated augmented matrix:

$$
\left[
  \begin{array}{rr|r}
    2 & 3 & 7 \\
    2 & -8 & 6 \\
  \end{array}
\right]
$$

We can use the following notation to describe this operation:

$$ Eq. (2) \leftarrow 2\times Eq. (2) $$

Note importantly that this does not change anything about what these pair of equations are telling us about the unknown solution vector $\pmb{x}$ which although it doesn't appear is implicilty defined by this augmented equation.

The next step was to subtract the first equation from the updated second ($ Eq. (2) \leftarrow Eq. (2) - Eq. (1) $):

$$
\left[
  \begin{array}{rr|r}
    2 & 3 & 7 \\
    0 & -11 & -1 \\
  \end{array}
\right]
$$

The square matrix that is now in the $A$ position of this augmented system is an example of an *upper-triangular* matrix - all entries below the diagonal are zero.  For such a matrix we can perform back substitution - starting at the bottom to solve trivially for the final unknown ($y$ here which clearly takes the value $-1/-11$), and then using this knowledge working our way up to solve for each remaining unknown in turn, here just $x$ (solving $2x + 3\times (1/11) = 7$).

Note that we can perform the similar substitution if we had a lower triangular matrix, first finding the first unknown and then working our way forward through the remaining unknows - hence in this case *forward substitution*.

Note that if we wished we could of course continue working on the augmented matrix to make the $A$ component diagonal: divide the second equation by 11 and multiply by 3 ($ Eq. (2) \leftarrow (3/11)\times Eq. (2) $) and add it to the first ($ Eq. (1) \leftarrow Eq. (1) +  Eq. (2) $):

$$
\left[
  \begin{array}{rr|r}
    2 & 0 & 7-3/11\\
    0 & -3 & -3/11 \\
  \end{array}
\right]
$$

and we can further make it the identity by dividing the rows by $2$ and $-3$ respectively ($ Eq. (1) \leftarrow (1/2)\times Eq. (1) $, $ Eq. (2) \leftarrow (-1/3)\times Eq. (2) $) :

$$
\left[
  \begin{array}{rr|r}
    1 & 0 & (7-3/11)/2 \\
    0 & 1 & 1/11 \\
  \end{array}
\right]
$$

Each of these augmented matrices encodes exactly the same information as the original matrix system in terms of the unknown vector $\pmb{x}$, and hence this is telling us that

$$ \pmb{x} = I \pmb{x} = \left[
  \begin{array}{c}
    (7-3/11)/2 \\
    1/11 \\
  \end{array}
\right]
$$

i.e. exactly the solution we found when we performed back substitution from the upper-triangular form of the augmented system.


### <span style="color:blue">Exercise 5.2: Gaussian elimination $3 \times 3$ example (by hand) </span>

Consider the system of linear equations

\begin{align*}
  2x + 3y - 4z &= 5 \\
  6x + 8y + 2z &= 3 \\
  4x + 8y - 6z &= 19
\end{align*}

write this in matrix form, form the corresponding augmented system and perform row operations until you get to upper-triangular form, find the solution using back substitution (**do this all with pen and paper**).

Write some code to check your answer using `numpy.dot(linalg.inv(A),b)`.

You should find $x=-6$, $y=5$, $z=-1/2$.


## Gaussian elimination - algorithm and code

Note that we are free to perform the following operations on the augmented system without changing the corresponding solution:

* Exchange two rows (refer to the section on *partial pivoting* next week)

* Multiply a row by a non-zero constant ($Eq. (i)\leftarrow \lambda \times Eq.(i)$)

* Subtracting a (non-zero) multiple of one row with another ($Eq. (i)\leftarrow Eq. (i) - \lambda \times Eq.(j)$)


Note that the equation/row being subtracted is termed the *pivot*.


Let's consider the algorithm mid-way working on an arbitrary matrix system, i.e. assume that the first $k$ rows (i.e. above the horizontal dashed line in the matrix below) have already been transformed into upper-triangular form, while the equations/rows below are not yet in this form.

The augmented equation in this case can be assumed to look like

$$
\left[
  \begin{array}{rrrrrrrrr|r}
    A_{11} & A_{12} & A_{13} & \cdots & A_{1k} & \cdots & A_{1j} & \cdots & A_{1n} & b_1 \\
    0      & A_{22} & A_{23} & \cdots & A_{2k} & \cdots & A_{2j} & \cdots & A_{2n} & b_2 \\
    0      & 0      & A_{33} & \cdots & A_{3k} & \cdots & A_{3j} & \cdots & A_{3n} & b_3 \\
    \vdots & \vdots & \vdots & \ddots & \vdots & \ddots & \vdots & \ddots & \vdots & \vdots \\
    0      & 0      & 0      & \cdots & A_{kk} & \cdots & A_{kj} & \cdots & A_{kn} & b_k \\    
\hdashline
    \vdots & \vdots & \vdots & \ddots & \vdots & \ddots & \vdots & \ddots & \vdots & \vdots \\
    0      & 0      & 0      & \cdots & A_{ik} & \cdots & A_{ij} & \cdots & A_{in} & b_i \\
    \vdots & \vdots & \vdots & \ddots & \vdots & \ddots & \vdots & \ddots & \vdots & \vdots \\
    0      & 0      & 0      & \cdots & A_{nk} & \cdots & A_{nj} & \cdots & A_{nn} & b_n \\
\end{array}
\right]
$$

Remember that here as we are mid-way through the algorithm the $A$'s and $b$'s in the above are not the same as in the original system!

Our aim as a next step in the algorithm is to use row $k$ (the pivot row) to *eliminate* $A_{ik}$, and we need to do this for all of the rows $i$ below the pivot, i.e. for all $i>k$.

To eliminate $A_{ik}$ for a single row $i$ we need to perform the operation 
$$ Eq. (i)\leftarrow Eq. (i) - \frac{A_{ik}}{A_{kk}} \times Eq.(k) $$

or equivalently

\begin{align}
A_{ij} &\leftarrow A_{ij} - \frac{A_{ik}}{A_{kk}} A_{kj}, \quad j=k,k+1,\ldots,n\\
b_i &\leftarrow b_i - \frac{A_{ik}}{A_{kk}} b_{k}
\end{align}

$j$ only needs to run from $k$ upwards as we can assume that the earlier entries in column $i$ have already been set to zero, and also that the corresponding terms from the pivot row are also zero (we don't need to perform operations that we know involve the addition of zeros!).

And to eliminate these entries for all rows below the pivot we need to repeat for all $i>k$.

### <span style="color:blue">Exercise 5.3: Gaussian elimination</span>

Write some code that takes a matrix $A$ and a vector $\pmb{b}$ and converts it into upper-triangular form using the above algorithm. For the $2 \times 2$ and $3\times 3$ examples from above compare the resulting $A$ and $\pmb{b}$ you obtain following elimination.


In [ ]:
def upper_triangle(A, b):
    # Assuming it is a square matrix of size nxn.
    n = numpy.size(b)
    
    # Loop over each pivot row.
    for k in range(n-1):
        # your code here

# Initialise a test matrix and RHS. Make sure they are initialised as floats.
A = # your code here
b = # your code here

upper_triangle(A, b)

# Here is a new trick for you - "pretty print"
from pprint import pprint
pprint(A)
pprint(b)

### Back substitution

Now that we have an augmented system in the upper-triangular form

$$
\left[
  \begin{array}{rrrrr|r}
    A_{11} & A_{12} & A_{13} & \cdots & A_{1n} &  b_1 \\
    0      & A_{22} & A_{23} & \cdots & A_{2n} &  b_2 \\
    0      & 0      & A_{33} & \cdots & A_{3n} &  b_3 \\
    \vdots & \vdots & \vdots & \ddots & \vdots & \vdots \\
    0      & 0      & 0      & \cdots & A_{nn} &  b_n \\    
\end{array}
\right]
$$

where the solution $\pmb{x}$ of the original system also satisfies $A\pmb{x}=\pmb{b}$ for the $A$ and $\pmb{b}$ in the above upper-triangular form (rather than the original $A$ and $\pmb{b}$!).

We can solve the final equation row to yield

$$x_n = \frac{b_n}{A_{nn}}$$

The second to last equation then yields (note we've introduced a comma in the subscripts here simply to make the more complex double indices easier to read)

\begin{align}
A_{n-1,n-1}x_{n-1} + A_{n-1,n}x_n &= b_{n-1}\\
\implies x_{n-1} = \frac{b_{n-1} - A_{n-1,n}x_n}{A_{n-1,n-1}}\\
\implies x_{n-1} = \frac{b_{n-1} - A_{n-1,n}\frac{b_n}{A_{nn}}}{A_{n-1,n-1}}
\end{align}

and so on to row $k$ which yields

\begin{align}
A_{k,k}x_{k} + A_{k,k+1}x_{k+1} +\cdots +  A_{k,n}x_n &= b_{k}\\
\iff A_{k,k}x_{k} + \sum_{j=k+1}^{n}A_{kj}x_j &= b_{k}\\
\implies x_{k} &= \left( b_k - \sum_{j=k+1}^{n}A_{kj}x_j\right)\frac{1}{A_{kk}}
\end{align}

### <span style="color:blue">Exercise 5.4: Back substitution</span>

Extend your code to perform back substitution and hence to obtain the final solution $\pmb{x}$.  Check against the solutions found earlier.  Come up with some random $n\times n$ matrices (you can use `numpy.random.rand` for that) and check your code against `numpy.dot(linalg.inv(A),b)` (remember to use the original $A$ and $\pmb{b}$ here of course!)


In [ ]:
# This function assumes that A is already an upper triangular matrix.
def back_substitution(A, b):
    n = numpy.size(b)
    
    x = numpy.zeros(n)
    for k in range(#your code here):
        #your code here
    
    return x

# and this A is the upper triangular matrix carried forward 
# from the Python box above
pprint(back_substitution(A, b)) # the solution using your codes
        
# Reinitialise A and b and check your answer against just 
# multiplying A inverse and b   
# write your code here


### Gauss-Jordan elimination

Recall that for the augmented matrix example we did by hand above we continued past the upper-triangular form so that the augmented matrix had the identity matrix in the $A$ location. This algorithm has the name Gauss-Jordan elimination but note that it requires more operations than the conversion to upper-triangular form followed by back subsitution and so is only of academic interest.

### Matrix inversion

Note that if we were to form the augmented equation with the full identity matrix in the place of the vector $\pmb{b}$, i.e. $[A|I]$ and performed row operations exactly as above until $A$ is transformed into the identity matrix $I$, then we would be left with the inverse of $A$ in the original $I$ location, i.e.

$$ [A|I] \rightarrow [I|A^{-1}] $$

### <span style="color:blue">Exercise 5.5: Matrix inversion</span>

Try updating your code to construct the inverse matrix. Check your answer against the inverse matrix given by a built-in function.

Hint: Once you have performed your Gaussian elimination to transform A into an upper triangular matrix, perform another elimination "from bottom to top" to transform A into a diagonal matrix.

In [ ]:
# Updated version of the upper_triangular function that
# assumes that a matrix, B, is in the old vector location
# in the augmented system, and applies the same operations to
# B as to A
def upper_triangle2(A, B):
    # your code here

                
# Function which transforms the matrix into lower triangular 
# form - the point here is that if you give it a 
# matrix that is already in upper triangular form, then the 
# result will be a diagonal matrix
def lower_triangle2(A, B):
    # your code here

    
# Let's redefine A as our matrix above
A = # your code here
# and B is the identity of the corresponding size
B = # your code here

# transform A into upper triangular form (and perform the same operations on B)
# your code here

# now make this updated A lower triangular as well (the result should be diagonal)
# your code here

# final step is just to divide each row through by the value of the diagonal
# to end up with the identity in the place of A
# your code here

# print final A and B and check solution
# your code here

### <span style="color:blue">Exercise 5.6: Zeros on the diagonal</span>

You may have noticed above that we have no way of guaranteeing that the $A_{kk}$ we divide through by in the Guassian elimination or back substitution algorithms is non-zero (or not very small which will also lead to computational problems).
Note also that we commented that we are free to exchange two rows in our augmented system - how could you use this fact to build robustness into our algorithms in order to deal with matrices for which our algorithms do lead to very small or zero $A_{kk}$ values?  

More on this next week!